src/application/embeddings/file_parser.py

```diff
@@ -0,0 +1,128 @@
+import io
+from logging import Logger
+from tempfile import TemporaryDirectory
+
+from typing import Generator
+from zipfile import BadZipFile, ZipFile
+from pymupdf import Document
+from fastapi import File, UploadFile
+
+from src.constants import MD_EXTENSION, PDF_EXTENSION, SUPPORTED_FILES_IN_ZIP_TUPLE, SUPPORTED_FILES_TUPLE, TEXT_EXTENSION, ZIP_EXTENSION
+from src.application.embeddings.errors import InvalidFileExtensionError
+
+
+class FileParser:
+    """
+    A utility class for parsing and converting different types of files to string format.
+    
+        This class provides methods to handle various file types including:
+        - Text files (.txt)
+        - Markdown files (.md)
+        - PDF files (.pdf)
+        - ZIP archives containing supported file types
+    
+        The parser can:
+        - Convert text and markdown files to strings
+        - Extract text content from PDF files
+        - Process ZIP files and extract supported file types within them
+        - Handle file content encoding and conversion
+    
+        Args:
+            logger (Logger): A logger instance for tracking operations and debugging    
+    """
+    
+    def __init__(self, logger: Logger):
+        self.logger = logger
+
+    def _convert_bytes_to_str(self, content: bytes) -> str:
+        return content.decode("utf-8")
+
+    def _convert_text_to_str(self, file: UploadFile = File(...)) -> str:
+        self.logger.debug(f"Converting text file {file.filename} to string...")
+        content = file.file.read()
+        return self._convert_bytes_to_str(content)
+
+    def _convert_from_doc_to_str(self, doc: Document) -> Generator[str, None, None]:
+        for page in doc:
+            yield page.get_text()
+
+    def _convert_pdf_to_str(self, file: UploadFile) -> Generator[str, None, None]:
+        self.logger.debug(f'Extracting text from PDF file {file.filename}')
+        file_content = file.file.read()
+        doc = Document(stream=file_content)
+        yield from self._convert_from_doc_to_str(doc)
+    
+    def _extract_documents_from_zip_file(self, file: UploadFile = File(...)) -> Generator[str, None, None]:
+        self.logger.debug(f'Extracting files from zip file {file.filename}')
+        try:
+            content = file.file.read()
+            with (
+                TemporaryDirectory() as temp_dir,
```

Comment:
how and where is this temp dir created? is it in /tmp? 

we will have to make sure when building the plugin (and creating it in the marketplace) that an empty dir is created for having write access to the fs if needed.

Reply:
also when is the folder removed?

Reply:
Regarding the access, the library ensures permissions are given to the process creating the variable. We could give a try on a built image to ensure everything works as intended.